spec/units/clever_tap_client_spec.rb

```diff
@@ -38,4 +39,272 @@
+      client.on_successful_upload(&success_proc)
+      client.upload([event1, event2])
+    end
+
+    before { setup_clevertap_client }
+
+    context 'when upload records are homogenous' do
+      context 'and valid records' do
+        let(:event1) { event_factory('1414', 'John') }
+        let(:event2) { event_factory('1515', 'Jill') }
+
+        context 'and objects fit `upload_limit`' do
+          VCR.use_cassette('upload within Event limit') do
+            it 'returns an array with one successful `Response` object' do
+              expect(subject.count).to eq 1
+              expect(subject.first.success).to be true
+              expect(subject.first.class).to eq CleverTap::Response
+            end
+
+            it_behaves_like 'it calls success hook', 'once'
+          end
+        end
+
+        context 'and objects do not fit `upload_limit`' do
+          before do
+            allow(CleverTap::Event)
+              .to receive(:upload_limit).and_return(1)
+          end
+
+          VCR.use_cassette('upload out of Event limit') do
+            it 'returns an array with two successful `Response` objects' do
+              expect(subject.count).to eq 2
+              expect(subject.all?(&:success)).to be true
+              expect(
+                subject.all? { |r| r.class == CleverTap::Response }
+              ).to be true
+            end
+
+            it_behaves_like 'it calls success hook', 'twice'
+          end
+        end
+      end
+
+      context 'and invalid records' do
+        let(:failure_proc) { proc { 'sample proc' } }
+
+        subject do
+          client = described_class.new
+          client.on_failed_upload(&failure_proc)
+          client.upload([profile])
+        end
+
+        let(:profile) { profile_factory }
+
+        it 'returns an array with one failed `Response` object' do
+          expect(subject.count).to eq 1
+          expect(subject.first.class).to eq CleverTap::Response
+          expect(subject.first.success).to be false
+        end
+
+        it_behaves_like 'it calls failure hook', 'once'
+      end
+    end
+
+    context 'when objects are not homogenous' do
+      let(:event1) { event_factory('1414', 'John') }
+      let(:event2) { {} }
+
+      it_behaves_like 'it raises an inconcistency error'
+    end
+  end
+
+  describe '#request_body' do
+    subject { described_class.new('123456', 'passcode') }
+    let(:records) do
+      [
+        { 'ID' => '123', 'Name' => 'John' },
+        { 'ID' => '456', 'Name' => 'Jill' }
+      ]
+    end
+
+    it 'converts records hash to json' do
+      expect(subject.send(:request_body, records))
+        .to eq({ 'd' => records }.to_json)
+    end
+  end
+
+  describe '#determine_type' do
+    subject { described_class.new('123456', 'passcode') }
+    let(:records) { [{}, []] }
+    let(:record) { 'sample' }
+
+    it 'returns the class of the first array element' do
+      expect(subject.send(:determine_type, records)).to eq Hash
```

Comment:
you don't need `send` here

Reply:
sure, I forgot to remove it after I (unsuccessfully) tested converting this part to a shared example